[Cocoa] colorSpace values missing from videoTrack.configuration with MP4s missing explicit color information

jernoble opened the pull request:
#### 41652ca1b8ff310d8f01acb27830c8ae31101145
<pre>
[Cocoa] colorSpace values missing from videoTrack.configuration with MP4s missing explicit color information
<a href="https://bugs.webkit.org/show_bug.cgi?id=257911">https://bugs.webkit.org/show_bug.cgi?id=257911</a>
rdar://110545901

Reviewed by Andy Estes.

Read the default colorspace values for a given codec and size from VideoToolbox.

* LayoutTests/media/track/video-track-configuration-mp4-default-colorspace-expected.txt: Added.
* LayoutTests/media/track/video-track-configuration-mp4-default-colorspace.html: Added.

Canonical link: <a href="https://commits.webkit.org/265053@main">https://commits.webkit.org/265053@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7606e44c1d61d3c36b8442a257315155b54e922

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9631 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9886 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10131 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11292 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9412 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9641 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11866 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9842 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12334 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9783 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10634 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8196 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11451 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7930 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8753 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16165 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9024 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8902 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12242 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9393 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7628 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8582 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/8657 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2316 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12803 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9167 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.cpp:
```diff
@@ -59,7 +60,28 @@ std::optional<PlatformVideoColorSpace> colorSpaceFromFormatDescription(CMFormatD
-    if (auto primaries = static_cast<CFStringRef>(PAL::CMFormatDescriptionGetExtension(formatDescription, PAL::get_CoreMedia_kCMFormatDescriptionExtension_ColorPrimaries()))) {
+    auto primaries = static_cast<CFStringRef>(PAL::CMFormatDescriptionGetExtension(formatDescription, PAL::get_CoreMedia_kCMFormatDescriptionExtension_ColorPrimaries()));
+    auto transfer = static_cast<CFStringRef>(PAL::CMFormatDescriptionGetExtension(formatDescription, PAL::get_CoreMedia_kCMFormatDescriptionExtension_TransferFunction()));
+    auto matrix = static_cast<CFStringRef>(PAL::CMFormatDescriptionGetExtension(formatDescription, PAL::get_CoreMedia_kCMFormatDescriptionExtension_YCbCrMatrix()));
```
Do we need to `dynamic_cf_cast<CFStringRef>` to be extra sure these are strings?

jernoble replied:
Ah, good point. These methods return a `CFPropertyListRef`, so we absolutely should verify that they are indeed CFStringRefs.

aestes commented on Source/WebCore/PAL/pal/cf/VideoToolboxSoftLink.h:
```diff
@@ -69,6 +69,8 @@ SOFT_LINK_FUNCTION_FOR_HEADER(PAL, VideoToolbox, VTCompressionSessionPrepareToEn
 #define VTCompressionSessionPrepareToEncodeFrames softLink_VideoToolbox_VTCompressionSessionPrepareToEncodeFrames
 SOFT_LINK_FUNCTION_FOR_HEADER(PAL, VideoToolbox, VTCompressionSessionInvalidate, void, (VTCompressionSessionRef session), (session))
 #define VTCompressionSessionInvalidate softLink_VideoToolbox_VTCompressionSessionInvalidate
+SOFT_LINK_FUNCTION_FOR_HEADER(PAL, VideoToolbox, VTGetDefaultColorAttributesWithHints, OSType, (OSType codecTypeHint, CFStringRef colorSpaceNameHint, size_t widthHint, size_t heightHint, CFStringRef* colorPrimariesOut, CFStringRef* transferFunctionOut, CFStringRef* yCbCrMatrixOut), (codecTypeHint, colorSpaceNameHint, widthHint, heightHint, colorPrimariesOut, transferFunctionOut, yCbCrMatrixOut))
```
Should the return type be `OSStatus` instead of `OSType`, like it is in the `SOFT_LINK_FUNCTION_FOR_SOURCE`?

jernoble replied:
Sure should

jernoble committed:
[Cocoa] colorSpace values missing from videoTrack.configuration with MP4s missing explicit color information
https://bugs.webkit.org/show_bug.cgi?id=257911
rdar://110545901

Reviewed by Andy Estes.

Read the default colorspace values for a given codec and size from VideoToolbox.

* LayoutTests/media/track/video-track-configuration-mp4-default-colorspace-expected.txt: Added.
* LayoutTests/media/track/video-track-configuration-mp4-default-colorspace.html: Added.

Canonical link: https://commits.webkit.org/265053@main
